browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -75,17 +176,18 @@ void OnBeforeURLRequestAdBlockTP(
   }
   DCHECK_NE(ctx->request_identifier, 0UL);
 
-  g_brave_browser_process->ad_block_service()->GetTaskRunner()
-      ->PostTaskAndReply(FROM_HERE,
-                         base::BindOnce(&ShouldBlockAdOnTaskRunner, ctx),
-                         base::BindOnce(&OnShouldBlockAdResult, next_callback,
-                                        ctx));
-}
+  scoped_refptr<base::SequencedTaskRunner> task_runner =
+      g_brave_browser_process->ad_block_service()->GetTaskRunner();
 
-int OnBeforeURLRequest_AdBlockTPPreWork(
-    const ResponseCallback& next_callback,
-    std::shared_ptr<BraveRequestInfo> ctx) {
+  base::OnceCallback<void(base::Optional<std::string>)> cname_callback =
+      base::BindOnce(&ShouldBlockAdWithOptionalCname, task_runner,
+                     next_callback, ctx);
+
+  new AdblockCnameResolveHostClient(std::move(cname_callback), ctx);
```

Comment:
do you have any ideas or measurements about performance implications of this?

Reply:
Not really, I don't had a good sense of how that can be done in this instance. I did look into writing some browsertests, but I couldn't see how to mock the DNS responses.

Happy to hear suggestions on this if you have any - but from manual testing, I couldn't notice any slowdown at all.

Reply:
We can start with measuring the delay manually (check the time before `ResolveHost` and then in `OnResult`) and reporting it to `UMA_HISTOGRAM_TIMES`. Then after some browsing this value can be checked on `brave://histograms`. Would be nice to test an android apk.

as for browser tests, I see some examples using `FakeHostResolver` - did you check that?

Reply:
hmm, I might need some more detailed instructions, this is all new to me.

Should histogram measurement code be left in and merged? Or just temporarily added for manual testing?

As for browser tests, I'm not sure of the best way to force shields to use the fake resolver. It looks like `chrome/browser/net/dns_probe_browsertest.cc` uses this to set up a `DnsProbeService` with a fake resolver, and I can't tell what the equivalent approach would look like for this PR.
```cpp
void DnsProbeBrowserTest::SetActiveBrowser(Browser* browser) {
  delaying_dns_probe_service_ = static_cast<DelayingDnsProbeService*>(
      DnsProbeServiceFactory::GetInstance()->SetTestingFactoryAndUse(
          browser->profile(),
          base::BindRepeating(
              &DelayingDnsProbeService::Create,
              base::BindRepeating(&DnsProbeBrowserTest::GetNetworkContext,
                                  base::Unretained(this)),
              base::BindRepeating(
                  &DnsProbeBrowserTest::GetDnsConfigChangeManager,
                  base::Unretained(this)))));
  // If currently watching a NetErrorTabHelper, stop doing so before start
  // watching another.
  if (monitored_tab_helper_) {
    monitored_tab_helper_->set_dns_probe_status_snoop_callback_for_testing(
        NetErrorTabHelper::DnsProbeStatusSnoopCallback());
  }
  active_browser_ = browser;
  monitored_tab_helper_ = NetErrorTabHelper::FromWebContents(
      active_browser_->tab_strip_model()->GetActiveWebContents());
  monitored_tab_helper_->set_dns_probe_status_snoop_callback_for_testing(
      BindRepeating(&DnsProbeBrowserTest::OnDnsProbeStatusSent,
                    Unretained(this)));
}
```

Reply:
> Should histogram measurement code be left in and merged?
yes, it can be useful in the future. Basically the whole Chromium codebase is full of histograms. 
I've added some related ones like `Brave.ProxyingURLLoader.TotalRequestTime` or `Brave.OnBeforeURLRequest_Handler`